### PR TITLE
boot,bootloader: extract kernel in makeBootable20RunMode 

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds kernel extraction to makeBootable20RunMode.
It cheats a bit right now and writes an UC16 compatible
boot environment. But by doing so the system should be able
to successfully go into "run" mode.

### CODE REVIEW

File: boot/boot.go [hunk 1]
```diff
@@ -433,7 +442,7 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
 	}
 	bl, err := bootloader.Find(filepath.Join(runMnt, "ubuntu-seed"), opts)
 	if err != nil {
-		return fmt.Errorf("internal error: cannot find bootloader: %v", err)
+		return fmt.Errorf("internal error: cannot find recovery system bootloader: %v", err)
 	}
 	blVars := map[string]string{
 		"snapd_recovery_mode": "run",
```
Comment (pedronis) [inline boot/boot.go]: this should be done last no?

File: boot/boot.go [hunk 2]
```diff
@@ -442,6 +451,37 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
 		return fmt.Errorf("cannot set recovery system environment: %v", err)
 	}
 
+	// get the ubuntu-boot grub and extrac the kernel there
+	opts = &bootloader.Options{
+		// TODO:UC20: we use "recovery: true" here because on
+		// the partition the file layout of ubuntu-boot looks
+		// the same as ubuntu-seed
+		Recovery:              true,
```
Comment (pedronis) [inline boot/boot.go]: this is confusing name now then. I need to look how it's used to propose a different one, but not a blocker.

File: boot/boot.go [hunk 3]
```diff
@@ -442,6 +451,37 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
...
+	if err != nil {
+		return err
+	}
+	// TODO:UC20: extract in the new "static" image way
+	if err := bl.ExtractKernelAssets(bootWith.Kernel, kernelf); err != nil {
+		return err
+	}
+	// TODO:UC20: no need to set these bootvars here anymore, in the
+	// new UC20 world the good/try kernels have static names
+	blVars = map[string]string{
```
Comment (pedronis) [inline boot/boot.go]: this needs a
// XXX using the UC16/18 grubenv style until we have a UC20 grub.cfg and corresponding snapd early boot code

File: boot/boot.go [hunk 4]
```diff
@@ -442,6 +451,37 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
...
+		return fmt.Errorf("internal error: cannot find run system bootloader: %v", err)
+	}
+	kernelf, err := snap.Open(bootWith.KernelPath)
+	if err != nil {
+		return err
+	}
+	// TODO:UC20: extract in the new "static" image way
+	if err := bl.ExtractKernelAssets(bootWith.Kernel, kernelf); err != nil {
+		return err
+	}
```
Comment (pedronis) [inline boot/boot.go]: we need also to copy kernel and base to ubuntu-data/var/lib/snapd/snaps

Comment (mvo5) [reply boot/boot.go]: Thanks! I added a (hopefully) uncontroversial PR for this in https://github.com/snapcore/snapd/pull/7918

**Comment by mvo5** [comment]
This is closed in favor of https://github.com/snapcore/snapd/pull/7931 - we will get to the kernel extraction again once we have a clear idea what this will look like.

**Pull request closed without merging**